### PR TITLE
Fix memory leak in fread

### LINKED ISSUE #3292: fread:  memory not being freed
I have found that reading data into a data.table will cause a memory leak. This seems to happen whether I use fread or read_fst(from the fst package). I suspect that this could even be an R or OS issue, however, I have not found any cases of this on Stack Overflow...
 
Steps to reproduce:
 
1. Download the sample file here:
```
download.file("https://mdodemodiag770.blob.core.windows.net/testcontainer/test.csv", "test.csv")
```

2. Create an R script as the following:

```
tst <- data.table::fread("tst.csv");
rm(tst);
gc()
```
 
3. Execute the R script and check the results on your OS. I am using Ubuntu 18.04.1. I see the following results of my R process when I run the top command (2.6 gbs in Virtual memory);
 
```
  PID   USER        PR VIRT         RES        SHR   S   %CUP %MEM TIME + COMMAND
15827 patrick+  20   0  2.625g 1.269g 0.013g S   0.0  1.2   0:12.93 rsession
 
```
--If you want to run in it with valgrind, you can use the following:
4.  Run the following valgrind command (change the source script to the location of your R script from step 2):
```
R -d "valgrind --show-leak-kinds=all" -e "source('/home/patrick.howerter/valgrind_test.R')"
 
```
Here is the end of the valgrind output.
5.
```
==533==
==533== HEAP SUMMARY:
==533==     in use at exit: 134,671,721 bytes in 19,047 blocks
==533==   total heap usage: 42,608 allocs, 23,561 frees, 1,113,301,337 bytes allocated
==533==
==533== LEAK SUMMARY:
==533==    definitely lost: 0 bytes in 0 blocks
==533==    indirectly lost: 0 bytes in 0 blocks
==533==      possibly lost: 4,320 bytes in 15 blocks
==533==    still reachable: 134,667,401
 
bytes in 19,032 blocks
==533==         suppressed: 0 bytes in 0 blocks
==533== Rerun with --leak-check=full to see details of leaked memory
==533==
==533== For counts of detected and suppressed errors, rerun with: -v
==533== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
 

```
 
sessionInfo()
```
R version 3.5.2 (2018-12-20)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 18.04.1 LTS

Matrix products: default
BLAS: /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.7.1
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.7.1

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8   
 [6] LC_MESSAGES=en_US.UTF-8    LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
[1] bit_1.1-14        compiler_3.5.2    tools_3.5.2       yaml_2.2.0        bit64_0.9-7       data.table_1.12.0

``

### PR DESCRIPTION
**Pull request by jimhester**
Fixes #3292

### CODE REVIEW

**Comment by jimhester** [comment]
Before this PR you can see the memory leak with the increase in Vcells

``` r
dummy <- rep("1\t2\t3\t4\t5", 10000000)
writeLines(dummy, "out.tsv")

invisible(data.table::fread("out.tsv"))
for (i in 1:10) {
  data.table::fread("out.tsv")
  print(gc())
}
#>            used  (Mb) gc trigger  (Mb) limit (Mb) max used  (Mb)
#> Ncells   539635  28.9     918942  49.1         NA   850004  45.4
#> Vcells 16041623 122.4   41582216 317.3      32768 41041748 313.2
#>            used  (Mb) gc trigger  (Mb) limit (Mb) max used  (Mb)
#> Ncells   539797  28.9     918942  49.1         NA   850004  45.4
#> Vcells 18540198 141.5   54339092 414.6      32768 43540348 332.2
#>            used  (Mb) gc trigger  (Mb) limit (Mb) max used  (Mb)
#> Ncells   539799  28.9     918942  49.1         NA   850004  45.4
#> Vcells 21040504 160.6   54339092 414.6      32768 46043083 351.3
#>            used  (Mb) gc trigger  (Mb) limit (Mb) max used  (Mb)
#> Ncells   539801  28.9     918942  49.1         NA   850004  45.4
#> Vcells 23540814 179.7   54339092 414.6      32768 48543401 370.4
#>            used  (Mb) gc trigger  (Mb) limit (Mb) max used  (Mb)
#> Ncells   539803  28.9     918942  49.1         NA   850004  45.4
#> Vcells 26041116 198.7   54339092 414.6      32768 51043727 389.5
#>            used  (Mb) gc trigger  (Mb) limit (Mb) max used  (Mb)
#> Ncells   539805  28.9     918942  49.1         NA   850004  45.4
#> Vcells 28541418 217.8   54339092 414.6      32768 53544029 408.6
#>            used  (Mb) gc trigger  (Mb) limit (Mb) max used  (Mb)
#> Ncells   539807  28.9     918942  49.1         NA   850004  45.4
#> Vcells 31041722 236.9   69340920 529.1      32768 56041872 427.6
#>            used  (Mb) gc trigger  (Mb) limit (Mb) max used  (Mb)
#> Ncells   539809  28.9     918942  49.1         NA   850004  45.4
#> Vcells 33542027 256.0   69340920 529.1      32768 58544647 446.7
#>            used  (Mb) gc trigger  (Mb) limit (Mb) max used  (Mb)
#> Ncells   539811  28.9     918942  49.1         NA   850004  45.4
#> Vcells 36042332 275.0   69340920 529.1      32768 61044961 465.8
#>            used  (Mb) gc trigger  (Mb) limit (Mb) max used  (Mb)
#> Ncells   539813  28.9     918942  49.1         NA   850004  45.4
#> Vcells 38542637 294.1   69340920 529.1      32768 63545275 484.9
print(gc())
#>            used  (Mb) gc trigger  (Mb) limit (Mb) max used  (Mb)
#> Ncells   539779  28.9     918942  49.1         NA   850004  45.4
#> Vcells 38540380 294.1   69340920 529.1      32768 63545275 484.9
```

<sup>Created on 2020-09-18 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

Afterwards the memory leak is gone and the Vcells are basically constant.
``` r
dummy <- rep("1\t2\t3\t4\t5", 10000000)
writeLines(dummy, "out.tsv")

invisible(data.table::fread("out.tsv"))
for (i in 1:10) {
  data.table::fread("out.tsv")
  print(gc())
}
#>            used (Mb) gc trigger  (Mb) limit (Mb) max used  (Mb)
#> Ncells   539635 28.9     918942  49.1         NA   850004  45.4
#> Vcells 11041081 84.3   52078987 397.4      32768 63993871 488.3
#>            used (Mb) gc trigger  (Mb) limit (Mb) max used  (Mb)
#> Ncells   539797 28.9     918942  49.1         NA   850004  45.4
#> Vcells 11039362 84.3   41663190 317.9      32768 63993871 488.3
#>            used (Mb) gc trigger  (Mb) limit (Mb) max used  (Mb)
#> Ncells   539799 28.9     918942  49.1         NA   850004  45.4
#> Vcells 11039382 84.3   52228899 398.5      32768 63993871 488.3
#>            used (Mb) gc trigger  (Mb) limit (Mb) max used  (Mb)
#> Ncells   539801 28.9     918942  49.1         NA   850004  45.4
#> Vcells 11039406 84.3   41783120 318.8      32768 63993871 488.3
#>            used (Mb) gc trigger  (Mb) limit (Mb) max used  (Mb)
#> Ncells   539803 28.9     918942  49.1         NA   850004  45.4
#> Vcells 11039422 84.3   52228944 398.5      32768 63993871 488.3
#>            used (Mb) gc trigger  (Mb) limit (Mb) max used  (Mb)
#> Ncells   539805 28.9     918942  49.1         NA   850004  45.4
#> Vcells 11039438 84.3   41783156 318.8      32768 63993871 488.3
#>            used (Mb) gc trigger  (Mb) limit (Mb) max used  (Mb)
#> Ncells   539807 28.9     918942  49.1         NA   850004  45.4
#> Vcells 11039456 84.3   52228984 398.5      32768 63993871 488.3
#>            used (Mb) gc trigger  (Mb) limit (Mb) max used  (Mb)
#> Ncells   539809 28.9     918942  49.1         NA   850004  45.4
#> Vcells 11039475 84.3   41783188 318.8      32768 63993871 488.3
#>            used (Mb) gc trigger  (Mb) limit (Mb) max used  (Mb)
#> Ncells   539811 28.9     918942  49.1         NA   850004  45.4
#> Vcells 11039494 84.3   52229028 398.5      32768 63993871 488.3
#>            used (Mb) gc trigger  (Mb) limit (Mb) max used  (Mb)
#> Ncells   539813 28.9     918942  49.1         NA   850004  45.4
#> Vcells 11039513 84.3   50203867 383.1      32768 63993871 488.3
print(gc())
#>            used (Mb) gc trigger  (Mb) limit (Mb) max used  (Mb)
#> Ncells   539779 28.9     918942  49.1         NA   850004  45.4
#> Vcells 11037405 84.3   40163094 306.5      32768 63993871 488.3
```

<sup>Created on 2020-09-18 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

This change does potentially perform a reallocation, whereas the previous code does not, but I am not sure there is any way around it.

File: NEWS.md [hunk 1]
```diff
@@ -12,6 +12,8 @@
 
 2. A regression in v1.13.0 resulted in installation on Mac often failing with `shared object 'datatable.so' not found`, and FreeBSD always failing with `expr: illegal option -- l`, [#4652](https://github.com/Rdatatable/data.table/issues/4652) [#4640](https://github.com/Rdatatable/data.table/issues/4640) [#4650](https://github.com/Rdatatable/data.table/issues/4650). Thanks to many for assistance including Simon Urbanek, Brian Ripley, Wes Morgan, and @ale07alvarez. There were no installation problems on Windows or Linux.
 
+3. A memory leak in `fread()` was fixed.
```
Comment (MichaelChirico) [inline NEWS.md:15]: please add a reference to the issue number and cite yourself and the issue author 😃  (c.f. bugfix 2. just above)

**Comment by jimhester** [comment]
An alternative implementation that also fixes the memory leak and avoids the reallocation would be to set the growable bit and the true length to be the allocated length. e.g. this diff from the current PR.

```diff
diff --git a/src/freadR.c b/src/freadR.c
index 8293586b..de260369 100644
--- a/src/freadR.c
+++ b/src/freadR.c
@@ -527,7 +527,9 @@ void setFinalNrow(size_t nrow) {
     if (nrow == dtnrows)
       return;
     for (int i=0; i<LENGTH(DT); i++) {
-      SET_VECTOR_ELT(DT, i, Rf_lengthgets(VECTOR_ELT(DT, i), nrow));
+      SETLENGTH(VECTOR_ELT(DT,i), nrow);
+      SET_TRUELENGTH(VECTOR_ELT(DT,i), dtnrow);
+      SET_GROWABLE_BIT(VECTOR_ELT(DT,i));
     }
   }
   R_FlushConsole(); // # 2481. Just a convenient place; nothing per se to do with setFinalNrow()
```

`SET_GROWABLE_BIT` is only available in R 3.4+, so this would need to be conditionally disabled for earlier R versions.

**Comment by jimhester** [comment]
I believe part that avoids the memory leak is actually setting the true length to `dtnrows` rather than `nrows`. Setting the growable bit is an optimization to let R use the extra memory.

**Comment by aitap** [comment]
Thank you very much for the response! I will try to investigate more in
this direction.

**Pull request merged**